### PR TITLE
enum is awesome!

### PR DESCRIPTION
**Pull request by t0lik**
What about using enums provided by TypeScript? This can lead to less errors (in this case there will be no more typo-errors in action and command names).

### CODE REVIEW

**Comment by luckyrat** [comment]
You experienced typo problems in action and command names? I found that the string literal types ensured that didn't cause any problems.

Now we've upgraded beyond Typescript 2.3 I'm happy to consider using new language features but don't typically want to make changes without a clear benefit beyond just it being fancy and new and fun to use.

Does the use of enums give any advantage over string literals in the current use cases?

**Comment by t0lik** [comment]
With enums you can simply don't remember exact name of command, just type in 'Commands.' and code completion helps you to select proper value from enum choices.
Second advantage: you wiil have the only one place with all the available commands so you can  control the whole set of commands much simplier.
Third advantage: you can describe each enum value to help understand future code readers what these values mean.
Ok. You can replace enums with a class with static public string fields. But this way (with class) is more prefferable when you have constants not related with each other.

File: background/commands.ts [hunk 1]
```diff
@@ -1,28 +1,34 @@
-class KFCommands {
 
+const enum Command {
```
Comment (luckyrat) [inline background/commands.ts]: Can we have this in a separate file (following the same approach as you used for the Actions)

File: common/Actions.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+const enum Actions {
+  CloseAllPanels = "closeAllPanels",
+  DetectForms = "detectForms",
+  GeneratePassword = "generatePassword",
+  ManualFill = "manualFill",
+  Primary = "primary",
+  RemoveSubmittedData = "removeSubmittedData",
+  ShowMatchedLoginsPanel = "showMatchedLoginsPanel",
+  SaveLatestLogin = "saveLatestLogin"
+}
```
Comment (luckyrat) [inline common/Actions.ts]: Can we give this a singular name instead please? `Action` instead of `Actions`

**Comment by t0lik** [comment]
Changes done! 
If we go further:
What do you think about using async/await instead of callbacks? I hate async callbacks. With async/await all the code looks much better and more readable and straitforward.

**Comment by luckyrat** [comment]
Thanks for the changes. They look good but I've run out of time tonight and would like to be able to test them a bit more thoroughly as part of merging them in so I'll have to postpone for a little bit.

Let's definitely keep this PR focussed on the enum change but yeah I'm happy to discuss using async/await in a follow-up PR. I had two reasons for not using this before: 

1) much of the code is converted from ES5 strict and refactoring all callbacks to use async/await would have introduced unnecessary risk 
2) I'm unsure how well the use of async/await works cross-browser - whatever we work towards should be able to work in Firefox, Chrome and Edge which all have subtly different interpretations of the "standard" WebExtensions API. It may well be perfectly fine to rewrite everything to async/await but I've not really looked into it in detail yet so can't say for sure.

Since it's probably going to be a lot of work to refactor everything, I suggest not diving in until we establish how well it will work in a cross-browser world. I also wouldn't want to accept a refactor that large onto the master branch until the first v2.0 release has had a little time to stabilise and starting a big refactor like this against a rapidly changing codebase can lead to a lot of merge conflict pain so we should be careful when we schedule it in order to limit that problem.

File: common/AddonMessage.ts [hunk 3]
```diff
@@ -17,7 +17,7 @@ interface AddonMessage {
     findMatches?: { uri: string };
     findMatchesResult?: any;
     logins?: keeLoginInfo[];
-    action?: "primary" | "detectForms" | "manualFill" | "generatePassword" | "closeAllPanels" | "showMatchedLoginsPanel" | "removeSubmittedData" | "saveLatestLogin";
+    action?: Action.Primary | Action.DetectForms | Action.ManualFill | Action.GeneratePassword | Action.CloseAllPanels | Action.ShowMatchedLoginsPanel | Action.RemoveSubmittedData | Action.SaveLatestLogin;
```
Comment (dlech) [inline common/AddonMessage.ts]: Could this simply be `action?: Action;`?

**Comment by t0lik** [comment]
You're quite right! Stable release has the highest priority! If I can help to find any errors you are welcome to invite me to this process!

File: common/AddonMessage.ts (hunk 3 above)
Comment (luckyrat) [reply common/AddonMessage.ts]: I agree with @dlech - that would be better.

**Comment by luckyrat** [comment]
This branch now has merge conflicts because of the changes I had to make yesterday to try to fix #27. I do intend to merge this into master once we can do so cleanly but I will wait for at least a week now - it's not going to have a noticeable benefit to the launch next week so I don't think it's worth us battling against any other urgent changes that need to be made and might add further conflicts.

**Comment by luckyrat** [comment]
Thanks for this.

I've just tried building without any of the extra entries in tslint.json and it works fine for me. Is that the same for you? If so, we should get rid of them rather than introduce further style inconsistencies down the line.

**Comment by t0lik** [comment]
I think it just works because of backward compatibility. I think if we have explicitly pointed new options we will easily can change them in future (if we wish to) and at the same time we don't need to remember their names and even to remember that those options exist at all.

P.S. I propose to replace comma operator that you used once in your code to make that for-loop more readable and clearer. And after that we can turn on option "ban-comma-operator"!

**Comment by luckyrat** [comment]
tslint has hundreds of options, with pretty decent documentation over at https://palantir.github.io/tslint/rules/ so I'm not sure we gain anything by explicitly listing them all in our config file unless we are planning to use them soon. In this case we could prefix it with //? like the other few config options we want to enable once the code has been tidied up a bit.

It looks like the other ones you've included are enabled anyway and they all seem like good things to enforce, apart from possibly jsdoc-format but only because I haven't had time to read up about what that actually does in practice.

Overall this seems like a nice step towards keeping the code quality high so thanks for that.

**Pull request merged**